core-rust/state-manager/src/staging/stage_tree.rs

```diff
@@ -62,243 +62,168 @@
  * permissions under this License.
  */
 
-use radix_engine::transaction::TransactionReceipt;
 use radix_engine::types::*;
-use radix_engine::{ledger::*, transaction::TransactionResult};
 
-use im::hashmap::HashMap as ImmutableHashMap;
 use sbor::rust::vec::Vec;
 use slotmap::{new_key_type, SlotMap};
 
-/// An immutable/persistent store (i.e a store built from a [`parent`] store
-/// shares data with it). Note: while from the abstract representation point
-/// of view you can delete keys, the underlying data is freed only and only
-/// if there are no references to it (there are no paths from any root reference
-/// to the node representing that (key, value)). For this reason, extra steps
-/// are taken to free up (key, value) pairs accumulating over time.
-/// This is intended as an wrapper/abstraction layer, so that changes to the
-/// ReadableSubstateStore/WriteableSubstateStore traits are easier to maintain.
-#[derive(Clone)]
-struct ImmutableStore {
-    outputs: ImmutableHashMap<SubstateId, OutputValue>,
-}
-
-impl ImmutableStore {
-    fn new() -> Self {
-        ImmutableStore {
-            outputs: ImmutableHashMap::new(),
-        }
-    }
-
-    fn from_parent(parent: &ImmutableStore) -> Self {
-        ImmutableStore {
-            // Note: this clone is O(1), only the root node is actually cloned
-            // Check im::collections::HashMap for details
-            outputs: parent.outputs.clone(),
-        }
-    }
+new_key_type! {
+    /// A key to a derived stage.
+    /// "Derived" here means "obtained by applying deltas to the root stage".
+    pub struct DerivedStageKey;
 }
 
-impl WriteableSubstateStore for ImmutableStore {
-    fn put_substate(&mut self, substate_id: SubstateId, output: OutputValue) {
-        self.outputs.insert(substate_id, output);
-    }
+/// A key to a (root or derived) stage.
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+pub enum StageKey {
+    Root,
+    Derived(DerivedStageKey),
 }
 
-impl ReadableSubstateStore for ImmutableStore {
-    fn get_substate(&self, substate_id: &SubstateId) -> Option<OutputValue> {
-        self.outputs.get(substate_id).cloned()
-    }
+/// A change to the accumulated state.
+pub trait Delta {
+    fn weight(&self) -> usize;
 }
 
-new_key_type! {
-    pub struct StagedSubstateStoreNodeKey;
+/// An accumulated state coming from multiple [`Delta`]s.
+pub trait Accumulator<D: Delta>: Default + Clone {
```

Comment:
I would add a comment to let the user know that `Clone` should be O(1) in order to keep the complexity as intended.

Reply:
Maybe rename it to `ImmutableAccumulator` to be more clear.

Reply:
I see why you are concerned, so I now did something even clearer 👍 PTAL.

_(I deliberately avoid using the word "immutable" here, because future developers [incl. future me] will definitely understand it wrong)_

Reply:
Thank you! That is better.